zkevm-circuits/src/tx_circuit.rs

```diff
@@ -73,20 +103,225 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
+                        tag.value_equals(TxFieldTag::CallData, Rotation::cur())(meta),
+                    ]),
+                ])
+            },
+            |meta| meta.query_advice(tx_table.value, Rotation::cur()),
+            |_| 0.expr(),
+        );
+        let tx_id_unchanged = IsEqualChip::configure(
+            meta,
+            |meta| {
+                and::expr(vec![
+                    meta.query_fixed(q_enable, Rotation::cur()),
+                    meta.query_advice(is_usable, Rotation::cur()),
+                ])
+            },
+            |meta| meta.query_advice(tx_table.tx_id, Rotation::cur()),
+            |meta| meta.query_advice(tx_table.tx_id, Rotation::next()),
+        );
+
+        let is_final = meta.advice_column();
+        let calldata_length = meta.advice_column();
+        let calldata_gas_cost_acc = meta.advice_column();
+        let chain_id = meta.advice_column();
+
+        let tx_sign_data_len = meta.advice_column();
+        let tx_sign_data_rlc = meta.advice_column();
+
+        Self::configure_lookups(
+            meta,
+            q_enable,
+            is_usable,
+            tag,
+            is_final,
+            calldata_length,
+            calldata_gas_cost_acc,
+            chain_id,
+            tx_sign_data_len,
+            tx_sign_data_rlc,
+            &value_is_zero,
+            tx_table.clone(),
+            keccak_table.clone(),
+            rlp_table,
+        );
 
         let sign_verify = SignVerifyConfig::new(meta, keccak_table.clone(), challenges);
 
+        meta.create_gate("tx call data bytes", |meta| {
+            let mut cb = BaseConstraintBuilder::default();
+
+            let is_final_cur = meta.query_advice(is_final, Rotation::cur());
+            cb.require_boolean("is_final is boolean", is_final_cur.clone());
+
+            // checks for any row, except the final call data byte.
+            cb.condition(not::expr(is_final_cur.clone()), |cb| {
+                cb.require_equal(
+                    "index::next == index::cur + 1",
+                    meta.query_advice(tx_table.index, Rotation::next()),
+                    meta.query_advice(tx_table.index, Rotation::cur()) + 1.expr(),
+                );
+                cb.require_equal(
+                    "tx_id::next == tx_id::cur",
+                    tx_id_unchanged.is_equal_expression.clone(),
+                    1.expr(),
+                );
+                cb.require_equal(
+                    "calldata_length::cur == calldata_length::next",
+                    meta.query_advice(calldata_length, Rotation::cur()),
+                    meta.query_advice(calldata_length, Rotation::next()),
+                );
+            });
+
+            // call data gas cost accumulator check.
+            cb.condition(
+                and::expr(vec![
+                    not::expr(is_final_cur.clone()),
+                    value_is_zero.is_equal_expression.clone(),
```

Comment:
This is a bug which can fail the unit test `tx_circuit_1tx` if we change the `CORRECT_MOCK_TXS[0].input` to has at least one zero bytes. To fix the bug, we should use value::next is zero as the condional parameter.